File: src/vector.cpp
#include "src.h"
#include <cmath>

Vector::Vector(double _x, double _y) {
  this->x = _x;
  this->y = _y;
}

Vector Vector::get() { return Vector(this->x, this->y); }

Vector &Vector::add(Vector &v) {
  this->x += v.x;
  this->y += v.y;
  return *this;
}

Vector &Vector::sub(Vector &v) {
  this->x -= v.x;
  this->y -= v.y;
  return *this;
}

Vector &Vector::multi(double n) {
  this->x *= n;
  this->y *= n;
  return *this;
}

Vector &Vector::div(double n) {
  this->x /= n;
  this->y /= n;
  return *this;
}

double Vector::mag() { return sqrt(pow(this->x, 2) + pow(this->y, 2)); }

Vector &Vector::normalize() {
  double m = this->mag();
  if (m != 0) {
    this->div(m);
  }
  return *this;
}

Vector &Vector::limit(double max) {
  if (this->mag() > max) {
    this->normalize();
    this->multi(max);
  }
  return *this;
}

// Static Methods
Vector Vector::AddVectors(Vector &v1, Vector &v2) {
  return Vector(v1.x + v2.x, v1.y + v2.y);
}
Vector Vector::SubVectors(Vector &v1, Vector &v2) {
  return Vector(v1.x - v2.x, v1.y - v2.y);
}
Vector Vector::Vector_multi(Vector &v1, double n) {
  return Vector(v1.x * n, v1.y * n);
}
Vector Vector::Vector_div(Vector &v1, double n) {
  return Vector(v1.x / n, v1.y / n);
}

File: go_src/vector.go
package src

// Vector Class and functions

import (
	"math"
)

type Vector struct {
	X float64
	Y float64
}

func NewVector(x, y float64) *Vector {
	return &Vector{X: x, Y: y}
}

func (v *Vector) Get() *Vector {
	vec := Vector{X: v.X, Y: v.Y}
	return &vec
}

func (v *Vector) Add(v2 Vector) *Vector {
	v.X += v2.X
	v.Y += v2.Y
	return v
}

func (v *Vector) Sub(v2 Vector) *Vector {
	v.X -= v2.X
	v.Y -= v2.Y
	return v
}

func (v *Vector) Multi(n float64) *Vector {
	v.X *= n
	v.Y *= n
	return v
}

func (v *Vector) Div(n float64) *Vector {
	v.X /= n
	v.Y /= n
	return v
}

func (v *Vector) Mag() float64 {
	return math.Sqrt(float64(v.X*v.X + v.Y*v.Y))
}

func (v *Vector) Normalize() *Vector {
	m := v.Mag()
	if m != 0 {
		v.Div(m)
	}
	return v
}

func (v *Vector) Limit(max float64) *Vector {
	if v.Mag() > max {
		v.Normalize()
		v.Multi(max)
	}
	return v
}

func AddVectors(v1, v2 Vector) *Vector {
	return &Vector{X: v1.X + v2.X, Y: v1.Y + v2.Y}
}

func SubVectors(v1, v2 Vector) *Vector {
	return &Vector{X: v1.X - v2.X, Y: v1.Y - v2.Y}
}

func Vector_Multi(v Vector, n float64) *Vector {
	return &Vector{X: v.X * n, Y: v.Y * n}
}

func Vector_Div(v Vector, n float64) *Vector {
	return &Vector{X: v.X / n, Y: v.Y / n}
}
